jordgubb/gfx.sy

```diff
@@ -0,0 +1,282 @@
+            love.graphics.setFont' f
+            drawTextFT' color, text, pos, alignemnt, maxWidth, 0., 1.
+        end,
+        drawTextFT: fn color, text, pos, alignemnt, maxWidth, r, s do
+            love.graphics.setFont' f
+            drawTextFT' color, text, pos, alignemnt, maxWidth, r, s
+        end,
+    }
+end
+
+Image :: blob {
+    img: *,
+    getWidth: fn -> float,
+    getHeight: fn -> float,
+    draw: fn (float, float), float, (float, float) -> void,
+    drawTinted: fn Color, (float, float), float, (float, float) -> void,
+    drawCenterd: fn (float, float), float, (float, float) -> void,
+    drawTintedCentered: fn Color, (float, float), float, (float, float) -> void,
+}
+
+drawImage :: fn tint, img: l.ImageData, pos, rot, scale do
+    setColor' tint
+    x :: love.graphics.draw
+    x' img, pos[0], pos[1], rot, scale[0], scale[1], 0., 0., 0.
+end
+
+drawImageCentered :: fn tint, img: l.ImageData, pos, rot, scale do
+    w :: img.getWidth' img
+    h :: img.getHeight' img
+    setColor' tint
+
+    // Work around for bug in typesystem, we don't copy
+    // functions if they're fields.
+    x :: love.graphics.draw
+    x' img, pos[0], pos[1], rot, scale[0], scale[1], w / 2., h / 2., 0.
+end
+
+newImage :: fn filepath: str -> Image
+    img : l.ImageData : love.graphics.newImage' filepath
+    Image {
+        img: img,
+        getWidth: fn -> img.getWidth' img end,
+        getHeight: fn -> img.getHeight' img end,
+        draw: fn pos, rot, scale do
+            drawImage' rgb(1., 1., 1.), img, pos, rot, scale
+        end,
+        drawTinted: fn tint, pos, rot, scale do
+            drawImage' tint, img, pos, rot, scale
+        end,
+        drawCenterd: fn pos, rot, scale do
+            drawImageCentered' rgb(1., 1., 1.), img, pos, rot, scale
+        end,
+        drawTintedCentered: fn tint, pos, rot, scale do
+            drawImageCentered' tint, img, pos, rot, scale
+        end,
+    }
+end
+
+SpriteSheet :: blob {
+    img: Image,
+    quads: [((int, int), (int, int))],
+    lquads: [l.Quad],
+
+    draw: fn Color, int, (float, float), float, (float, float) -> void,
+    drawCenterd: fn Color, int, (float, float), float, (float, float) -> void,
+}
+
+// TODO(ed): Raise to std-lib - maybe solve the 0 situation.
+for_range :: fn lo: *a, hi: *a, step: *a, f: fn *a -> void do
+    // NOTE(ed): Wont work for zero... :(
+    curr := lo
+    loop curr < hi do
+        f' curr
+        curr += step
+    end
+end
+
+// (w, h) - all in pixels
+newImplicitSpriteSheet :: fn img: Image, spriteDim: (int, int) ->
+    w :: spriteDim[0]
+    h :: spriteDim[1]
+    quads :: []
+    // TODO(ed): Maybe remove the remained so we only get full tiles?
+    // TODO(ed): Maybe add a divisibility test here?
+    for_range' 0, as_int(img.getWidth()), w, fn x -> do
+        for_range' 0, as_int(img.getHeight()), h, fn y -> do
+            quads -> push' ((x, y), (w, h))
+        end
+    end
```

Comment:
remember when we couldn't have for loops in other for loops? good times. (i had to do a double take here xd)